node/src/main/scala/coop/rchain/node/configuration/Configuration.scala

```diff
@@ -2,264 +2,124 @@ package coop.rchain.node.configuration
+import kamon.Kamon
+import kamon.system.SystemMetrics
+import kamon.zipkin.ZipkinReporter
 import monix.eval.Task
+import pureconfig._
+import pureconfig.generic.ProductHint
+import pureconfig.generic.auto._
 
+/**
+  * Configuration for RChain node instance
+  */
 object Configuration {
 
-  private val dockerProfile =
-    Profile("docker", dataDir = (() => Paths.get("/var/lib/rnode"), "Defaults to /var/lib/rnode"))
-
-  private val defaultProfile =
-    Profile(
-      "default",
-      dataDir = (() => Paths.get(sys.props("user.home"), ".rnode"), "Defaults to $HOME/.rnode")
+  /**
+    * Builds Configuration instance from CLI options.
+    * If config file is provided as part of CLI options, it shall be parsed and merged
+    * with CLI options having higher priority.
+    * @param options CLI options
+    * @return Configuration instance
+    */
+  def build(options: commandline.Options): (NodeConf, Config) = {
+    val profile = options.profile.toOption.flatMap(profiles.get).getOrElse(defaultProfile)
+    val dataDir = options.run.dataDir.getOrElse(profile.dataDir._1())
+    val configFile = options.run.configFile
+      .getOrElse(
+        dataDir.resolve("rnode.conf")
+      )
+      .toFile
+
+    /**
+      * Target Configuration is a merge of 3 sources according to the following priority, highest to lowers
+      * 1. optionsConfig - CLI options
+      * 2. fileConfig - `.conf` file provided via CLI option `-c`
+      * 3. defaultConfig - defaults in `node/src/main/resources/defaults.conf`,
+      *    defaultConfig is modified with parameters set in profile
+      */
+    val optionsConfig = ConfigSource.fromConfig(
+      ConfigMapper.fromOptions(options)
     )
+    val fileConfig =
+      if (configFile.exists())
+        ConfigSource.file(configFile)
+      else
+        ConfigSource.empty
+    val defaultConfig = ConfigSource
+      .resources("defaults.conf")
+      .withFallback(
+        ConfigSource.string(
+          s"default-data-dir = $dataDir"
+        )
+      )
 
-  private val profiles: Map[String, Profile] =
-    Map(defaultProfile.name -> defaultProfile, dockerProfile.name -> dockerProfile)
-
-  // within range HTTP2 RFC 7540
-  private val MaxMessageSizeMinimumValue: Int = 256 * 1024 // 0.25 MB
-  private val MaxMessageSizeMaximumValue: Int = 10 * 1024 * 1024
-
-  private val DataDirConfigKey     = s"${hocon.Server.Key}.${hocon.Server.keys.DataDir}"
-  private val CertificateConfigKey = s"${hocon.Tls.Key}.${hocon.Tls.keys.Certificate}"
-  private val KeyConfigKey         = s"${hocon.Tls.Key}.${hocon.Tls.keys.Key}"
-  private val GenesisPathConfigKey = s"${hocon.Casper.Key}.${hocon.Casper.keys.GenesisPath}"
-
-  def build(arguments: Seq[String]): Configuration = {
-    val configurationE = Try(buildConfiguration(arguments)).toEither
+    // Throw an error is unknown keys found
+    //implicit val hint = ProductHint[NodeConf](allowUnknownKeys = false)
```

Comment:
It will just pass through it without any effect. Thats  bad but we need to support unknown keys to store Kamon config on rnode.conf.
Though I was thinking that we can add one more conf file and name it `kamon.conf`. From one standpoint it should be good because that config is not determined by us. But we need additional config file in that case for configuring metrics.

Reply:
I enabled strict check and moved kamon conf to separate file. Path for it is not configurable,it should be next to `rnode.conf` and called `kamon.conf` https://github.com/rchain/rchain/pull/2912/commits/67e60ab94ee993659a10910f140c0715799d5981. Now if there are unknown key node won't start and throw an error.

Reply:
I agree to have `kamon.conf` in a separate file.

Is it possible to later add CLI flag to disable strict check? E.g. editor plugins supports multiple node versions and tracking what key is correct in which version can be challenging.

Reply:
But if user have some key in configuration - he's expecting for it to take an effect. And if we don't rise an error for unknown keys, it might be confusing.

Reply:
This is true is you are assuming that human is running the node, but if you want to write program that control the node it can be much more difficult to manage configuration.
E.g. when `--allow-private-addresses` was added to RNode, all previous versions would refuse to start with this parameter although there is no real problem.